uxarray/grid.py

```diff
@@ -0,0 +1,272 @@
+        print("initializing with gridspec")
+
+    # load mesh from a file
+    def from_file(self):
+        """Loads a mesh file Also, called by __init__ routine This routine will
+        automatically detect if it is a UGrid, SCRIP, Exodus, or shape file.
+
+        Raises:
+            RuntimeError: Invalid file type
+        """
+        # find the file type
+        self.meshFileType = self.find_type()
+
+        # call reader as per meshFileType
+        if self.meshFileType == "exo":
+            self.read_exodus(self.ext_ds)
+        elif self.meshFileType == "scrip":
+            self.read_scrip(self.ext_ds)
+        elif self.meshFileType == "ugrid":
+            self.read_ugrid(self.filepath)
+        elif self.meshFileType == "shp":
+            self.read_shpfile(self.filepath)
+        else:
+            raise RuntimeError("unknown file format.")
+
+    # helper function to find file type
+    def find_type(self):
+        """Checks file path and contents to determine file type Also, called by
+        __init__ routine This routine will automatically detect if it is a
+        UGrid, SCRIP, Exodus, or shape file.
+
+        Raises:
+            RuntimeError: Invalid file type
+        """
+
+        try:
+            # extract the file name and extension
+            path = PurePath(self.filepath)
+            file_extension = path.suffix
+
+            # try to open file with xarray
+            self.ext_ds = xr.open_dataset(self.filepath, mask_and_scale=False)
+            #
+        except (TypeError, AttributeError) as e:
+            msg = str(e) + ': {}'.format(self.filepath)
+            print(msg)
+            raise RuntimeError(msg)
+            exit
+        except (RuntimeError, OSError) as e:
+            # check if this is a shp file
+            # we won't use xarray to load that file
+            if file_extension == ".shp":
+                self.meshFileType = "shp"
+            else:
+                msg = str(e) + ': {}'.format(self.filepath)
+                print(msg)
+                raise RuntimeError(msg)
+                exit
+        except ValueError as e:
+            # check if this is a shp file
+            # we won't use xarray to load that file
+            if file_extension == ".shp":
+                self.meshFileType = "shp"
+            else:
+                msg = str(e) + ': {}'.format(self.filepath)
+                print(msg)
+                raise RuntimeError(msg)
+                exit
+
+        # Detect mesh file type, based on attributes of ext_ds:
+        # if ext_ds has coordx or coord - call it exo format
+        # if ext_ds has grid_size - call it SCRIP format
+        # if ext_ds has ? read as shape file populate_scrip_dataformat
+        # TODO: add detection of shpfile etc.
+        try:
+            self.ext_ds.coordx
+            self.meshFileType = "exo"
+        except AttributeError as e:
+            pass
+        try:
+            self.ext_ds.grid_center_lon
```

Comment:
Same as above

Reply:
same, this just checks for "grid_center_lon" presence in ext_ds